src/System.Text.Utf8/tests/RandomTests.cs

```diff
@@ -55,210 +56,107 @@ private static string GetStringLiteral(string s)
-                Assert.Equal(equal, s1FromSpan == s2);
-
-                Assert.Equal(equal, s1 == s2FromBytes);
-                Assert.Equal(equal, s1 == s2FromSpan);
-
-
-                Assert.Equal(equal, s1FromBytes.Equals(s2FromBytes));
-                Assert.Equal(equal, s1FromBytes.Equals(s2FromSpan));
-                Assert.Equal(equal, s1FromBytes.Equals(s2));
-
-                Assert.Equal(equal, s1FromSpan.Equals(s2FromBytes));
-                Assert.Equal(equal, s1FromSpan.Equals(s2FromSpan));
-                Assert.Equal(equal, s1FromSpan.Equals(s2));
-
-                Assert.Equal(equal, s1.EqualsUtf8String(s2FromBytes));
-                Assert.Equal(equal, s1.EqualsUtf8String(s2FromSpan));
-            }
+        public static object[][] LengthTestCases = {
+            new object[] { 0, default(Utf8String) },
+            new object[] { 0, Utf8String.Empty },
+            new object[] { 0, ""u8 },
+            new object[] { 4, "1258"u8 },
+            new object[] { 3, "\uABCD"u8 },
+            new object[] { 3, "\uABEE"u8 },
+            new object[] { 4, "a\uABEE"u8 },
+            new object[] { 5, "a\uABEEa"u8 },
+            new object[] { 8, "a\uABEE\uABCDa"u8 }
+        };
+        [Theory, MemberData("LengthTestCases")]
+        public void Length(int expectedLength, Utf8String s)
+        {
+            Assert.Equal(expectedLength, s.Length);
+        }
+
+        public static object[][] LengthInCodePointsTestCases = {
+            new object[] { 0, default(Utf8String) },
+            new object[] { 0, Utf8String.Empty },
+            new object[] { 0, ""u8 },
+            new object[] { 4, "1258"u8 },
+            new object[] { 1, "\uABCD"u8 },
+            new object[] { 1, "\uABEE"u8 },
+            new object[] { 2, "a\uABEE"u8 },
+            new object[] { 3, "a\uABEEa"u8 },
+            new object[] { 4, "a\uABEE\uABCDa"u8 }
+        };
+        [Theory, MemberData("LengthInCodePointsTestCases")]
+        public void LengthInCodePoints(int expectedLength, Utf8String s)
+        {
+            Assert.Equal(expectedLength, s.CodePoints.Count());
+        }
+
+        public static object[][] ToStringTestCases = {
+            new object[] { "", default(Utf8String) },
+            new object[] { "", Utf8String.Empty },
+            new object[] { "", ""u8 },
+            new object[] { "1258", "1258"u8 },
+            new object[] { "\uABCD", "\uABCD"u8 },
+            new object[] { "\uABEE", "\uABEE"u8 },
+            new object[] { "a\uABEE", "a\uABEE"u8 },
+            new object[] { "a\uABEEa", "a\uABEEa"u8 },
+            new object[] { "a\uABEE\uABCDa", "a\uABEE\uABCDa"u8 }
+        };
+        [Theory, MemberData("ToStringTestCases")]
+        public void ToString(string expected, Utf8String s)
+        {
+            Assert.Equal(expected, s.ToString());
+        }
+
+        public static object[][] StringEqualsTestCases = new object[][] {
+            // empty strings consistency
```

Comment:
I actually like breaking code with many test cases in the same place. When I'm breaking String.Equals I'm breaking String.Equals, I don't like writing too many comments just for the sake of explaining it where you can look at this simple 20 test cases and get conclusions yourself. Feels like breaking it would just make adding more test cases harder as I would have to search for different scenarios and had to understand the complexity of test infra while on the other hand I have just simple adding a dumb test case